language/move-vm/runtime/src/data_cache.rs

```diff
@@ -107,54 +97,57 @@ impl<'r, 'l, R: RemoteCache> TransactionDataCache<'r, 'l, R> {
+            let mut modules = BTreeMap::new();
+            for (module_name, module_blob) in account_data_cache.module_map {
+                modules.insert(module_name, Some(module_blob));
+            }
+
+            let mut resources = BTreeMap::new();
+            for (ty, (layout, gv)) in account_data_cache.data_map {
                 match gv.into_effect()? {
                     GlobalValueEffect::None => (),
                     GlobalValueEffect::Deleted => {
-                        if let TypeTag::Struct(s_tag) = self.loader.type_to_type_tag(&ty)? {
-                            vals.push((s_tag, None))
-                        } else {
-                            // non-struct top-level value; can't happen
-                            return Err(PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR));
-                        }
+                        let struct_tag = match self.loader.type_to_type_tag(&ty)? {
+                            TypeTag::Struct(struct_tag) => struct_tag,
+                            _ => return Err(PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR)),
+                        };
+                        resources.insert(struct_tag, None);
                     }
                     GlobalValueEffect::Changed(val) => {
-                        if let TypeTag::Struct(s_tag) = self.loader.type_to_type_tag(&ty)? {
-                            vals.push((s_tag, Some((ty_layout, val))))
-                        } else {
-                            // non-struct top-level value; can't happen
-                            return Err(PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR));
-                        }
+                        let struct_tag = match self.loader.type_to_type_tag(&ty)? {
+                            TypeTag::Struct(struct_tag) => struct_tag,
+                            _ => return Err(PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR)),
+                        };
+                        let resource_blob = val
+                            .simple_serialize(&layout)
+                            .ok_or_else(|| PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR))?;
+                        resources.insert(struct_tag, Some(resource_blob));
                     }
                 }
             }
-            if !vals.is_empty() {
-                resources.push((addr, vals));
-            }
-            modules.extend(
-                account_cache
-                    .module_map
-                    .into_iter()
-                    .map(|(module_id, blob)| (module_id, blob)),
-            );
+
+            account_changesets.insert(addr, AccountChangeSet { modules, resources });
         }
 
         let mut events = vec![];
         for (guid, seq_num, ty, ty_layout, val) in self.event_data {
             let ty_tag = self.loader.type_to_type_tag(&ty)?;
-            events.push((guid, seq_num, ty_tag, ty_layout, val))
+            // REVIEW: This error doesn't look right.
```

Comment:
I think this is right? You'll hit this if you do `emit_event<T>(msg : T)`, but `msg` doesn't match the layout of `T` (which should never happen)

Reply:
Yeah we discussed this last time. I just forgot to remove the comment.